controllers/clustertopology_controller_compute_test.go

```diff
@@ -250,7 +252,281 @@ func TestGetClass(t *testing.T) {
+				infrastructureCluster: infraCluster,
+				machineDeployments:    emptyMachineDeployments,
+			},
+		},
+		{
+			name:    "Cluster with infrastructure reference and control plane reference, no control plane infrastructure and a clusterClass with an infra requirement",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withControlPlaneRef(objToRef(controlPlane)).withInfraRef(objToRef(infraCluster)).Obj(),
+			class:   *clusterClassWithControlPlaneInfra,
+			objects: []client.Object{
+				controlPlane,
+				infraCluster,
+				clusterClassWithControlPlaneInfra,
+			},
+			// Expecting error from control plane having no valid control plane infrastructure with clusterclass requiring control plane infrastructure.
+			wantErr: true,
+		},
+		{
+			name:    "Cluster with control plane reference and with control plane infrastructure, but no cluster infrastructure",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withControlPlaneRef(objToRef(controlPlaneWithInfra)).Obj(),
+			objects: []client.Object{
+				controlPlaneWithInfra,
+				controlPlaneInfrastructureMachineTemplate,
+			},
+			// Expecting valid return with valid control plane state, but no control plane infrastructure, cluster infrastructure or machine deployment state defined.
+			want: &clusterTopologyState{
+				cluster:               newFakeCluster(metav1.NamespaceDefault, "cluster1").withControlPlaneRef(objToRef(controlPlaneWithInfra)).Obj(),
+				controlPlane:          controlPlaneTopologyState{object: controlPlaneWithInfra, infrastructureMachineTemplate: nil},
+				infrastructureCluster: nil,
+				machineDeployments:    emptyMachineDeployments,
+			},
+		},
+		{
+			name:    "Cluster with cluster infrastructure reference control plane reference and control plane infrastructure",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withInfraRef(objToRef(infraCluster)).withControlPlaneRef(objToRef(controlPlaneWithInfra)).Obj(),
+			class:   *clusterClassWithControlPlaneInfra,
+			objects: []client.Object{
+				infraCluster,
+				clusterClassWithControlPlaneInfra,
+				controlPlaneInfrastructureMachineTemplate,
+				controlPlaneWithInfra,
+			},
+			// Expecting valid return with valid control plane, control plane infrastructure and cluster infrastructure state, but no defined machine deployment state.
+			want: &clusterTopologyState{
+				cluster:               newFakeCluster(metav1.NamespaceDefault, "cluster1").withInfraRef(objToRef(infraCluster)).withControlPlaneRef(objToRef(controlPlaneWithInfra)).Obj(),
+				controlPlane:          controlPlaneTopologyState{object: controlPlaneWithInfra, infrastructureMachineTemplate: controlPlaneInfrastructureMachineTemplate},
+				infrastructureCluster: infraCluster,
+				machineDeployments:    emptyMachineDeployments,
+			},
+		},
+		{
+			name:    "Cluster with machine deployment state but no other states defined",
```

Comment:
Note: this in practice can't happen , because a cluster infrastructure and a control plane are required before creating workers

Reply:
Agreed - but I thought it was good to show all of the parts can work independently here.